gonat_linux.go

```diff
@@ -0,0 +1,337 @@
+	udpSocket          io.ReadWriteCloser
+	downstream         io.ReadWriter
+	opts               *Opts
+	bufferPool         BufferPool
+	ctrack             *ct.Conn
+	ctTimeout          uint32
+	randomPortSequence []uint16
+	portIndexes        map[uint8]map[Addr]int
+	connsByDownFT      map[FiveTuple]*conn
+	connsByUpFT        map[FiveTuple]*conn
+	fromDownstream     chan *IPPacket
+	toDownstream       chan *IPPacket
+	fromUpstream       chan *IPPacket
+	closedConns        chan *conn
+	close              chan interface{}
+}
+
+// NewServer constructs a new Server that reads packets from downstream
+// and writes response packets back to downstream.
+func NewServer(downstream io.ReadWriter, opts *Opts) (Server, error) {
+	err := opts.ApplyDefaults()
+	if err != nil {
+		return nil, errors.New("Error applying default options: %v", err)
+	}
+
+	log.Debugf("Outbound packets will use %v", opts.IFAddr)
+
+	ctrack, err := ct.Dial(nil)
+	if err != nil {
+		return nil, errors.New("Unable to obtain connection for managing conntrack: %v", err)
+	}
+	_ctTimeout := opts.IdleTimeout * 2
+	if _ctTimeout < MinConntrackTimeout {
+		_ctTimeout = MinConntrackTimeout
+	}
+	ctTimeout := uint32(_ctTimeout.Seconds())
+
+	// We create a random order for assigning new ports to minimize the chance of colliding
+	// with other running gonat instances.
+	randomPortSequence := make([]uint16, numEphemeralPorts)
+	for i := uint16(0); i < uint16(numEphemeralPorts); i++ {
+		randomPortSequence[i] = minEphemeralPort + i
+	}
+	rnd := rand.New(rand.NewSource(time.Now().UnixNano()))
+	rnd.Shuffle(numEphemeralPorts, func(i int, j int) {
+		randomPortSequence[i], randomPortSequence[j] = randomPortSequence[j], randomPortSequence[i]
+	})
+
+	s := &server{
+		downstream:         downstream,
+		opts:               opts,
+		bufferPool:         opts.BufferPool,
+		ctrack:             ctrack,
+		ctTimeout:          ctTimeout,
+		randomPortSequence: randomPortSequence,
+		portIndexes:        make(map[uint8]map[Addr]int),
+		connsByDownFT:      make(map[FiveTuple]*conn),
+		connsByUpFT:        make(map[FiveTuple]*conn),
+		fromDownstream:     make(chan *IPPacket, opts.BufferDepth),
+		toDownstream:       make(chan *IPPacket, opts.BufferDepth),
+		fromUpstream:       make(chan *IPPacket, opts.BufferDepth),
+		closedConns:        make(chan *conn, opts.BufferDepth),
+		close:              make(chan interface{}),
+	}
+	return s, nil
+}
+
+func (s *server) Serve() error {
+	var err error
+	s.tcpSocket, err = createSocket(FiveTuple{IPProto: syscall.IPPROTO_TCP, Src: Addr{s.opts.IFAddr, 0}})
+	if err != nil {
+		return err
+	}
+	ops.Go(func() { s.readFromUpstream(s.tcpSocket) })
+
+	s.udpSocket, err = createSocket(FiveTuple{IPProto: syscall.IPPROTO_UDP, Src: Addr{s.opts.IFAddr, 0}})
+	if err != nil {
+		return err
+	}
+	ops.Go(func() { s.readFromUpstream(s.udpSocket) })
+
+	ops.Go(s.trackStats)
+	ops.Go(s.dispatch)
+	ops.Go(s.writeToDownstream)
+	return s.readFromDownstream()
```

Comment:
It seems like the logic here is all single-threaded.  We do have separate routines reading and writing from downstream clients, as well as reading and writing to origins upstream.  They all seem to be (potentially) bottle-necked by dispatch though.  Dispatch does a lot too - searching for ports, opening sockets, calculating checksums.  The read/write methods on the other hand can just go as quickly as the TUN device allows - probably darn fast.  If they outpace dispatch, I think they'd just fill the buffers then start dropping packets, right?

This is pre-optimization talk, but if we start seeing lots of dropped packets, this could be a place to look.

Reply:
Yeah, the alternative to dispatch is to use locking. I intentionally keep the work in dispatch to a minimum for performance reasons. I find that having this kind of event loop goroutine makes it easier to reason about concurrency than a bunch of separate goroutines with locking.

> The read/write methods on the other hand can just go as quickly as the TUN device allows - probably darn fast

If the TUN device is like most network devices, it buffers internally and has a limit to have far it will buffer, at which point it back-pressures. Also, in production, the downstream end is not a TUN device but a Lantern client connected over the Internet via TCP with HTTPS, Lampshade, etc. If anything is going to cause dropped packets, it'll most likely be backups in writing to the client via the network.

I've done manual testing downloading 1 GB files as well as just general browsing, and I saw relatively few dropped packets. Still, worth keeping an eye on :)

Reply:
To help us understand if this becomes a problem in practice, I'm now distinguishing between invalid and dropped packets. Dropped should ideally remain at 0.

See https://github.com/getlantern/gonat/pull/1/commits/4e1c84eca4f7ba82d267bb5d699614b9d1780923

Reply:
Okay, thanks for the explanation!  Yeah, it would definitely increase the complexity to parallelize dispatch given all the memory we'd have to share.  Also, after looking at the code again, both reading routines (from up/downstream) will block when the buffers fill.  They won't just start dropping packets like I originally said.  So I think this is a total non-issue.  Still, your explanation was helpful for putting this library in context.

Reply:
Oh I missed your comment about distinguishing between invalid and dropped packets.  That seems pretty useful regardless of whether this particular place in the code causes any issues.